### ISSUE TITLE
[Bug]: Auto-scale nodes and LB survive after terraform destroy

### ISSUE THREAD

**Issue by valkenburg-prevue-ch** [kube-hetzner/terraform-hcloud-kube-hetzner#579]
### Description

Nodes created by the cluster's autoscaler are not part of the terraform state, and hence do not get destroyed when one destroys a cluster.

Is there a possibility to add a script which runs on destroy which removes all things created by the autoscaler?

### Kube.tf file

```terraform
autoscaler_nodepools = [
    {
      name        = "autoscaler"
      server_type = "cx31" # must be same or better than the control_plane server type (regarding disk size)!
      location    = "fsn1"
      min_nodes   = 0
      max_nodes   = 5
    }
  ]
```

### Platform

Linux

**Comment by mysticaltech** [comment]
@valkenburg-prevue-ch You are right. It's the same problem that we have with the ingress LB.

IMHO, requiring the `hcloud` cli could be a quick and dirty solution, but what would be best is to do the deletion via the [Hetzner Cloud API](https://docs.hetzner.cloud/) instead, as the autoscaled nodes are labeled already, so they are easy to select (as done below).

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/d179e3a32e6d7fb6fe8eb76d808074b4e1466d93/autoscaler-agents.tf#L105

So basically, we could have a local-exec that runs upon the last stages of the destroy, and that takes down these nodes via the API and using the token. And the same for the LB.

**Comment by valkenburg-prevue-ch** [comment]
I see the autoscaler is currently installed from 1 yaml file with 6 pages. It is straightforward to make that a helm chart, and install it as such using the terraform helm provider. This would allow you to use a [hook](https://helm.sh/docs/topics/charts_hooks/) which runs a job upon deletion. So you could use the autoscaler image and delete everything it created. Would that be an idea?

**Comment by mysticaltech** [comment]
@valkenburg-prevue-ch It could work I agree, but you're going into Kubernetes and that's slow and sometimes can get stuck. IMHO, we bypass all of this by taking the nodes down at the infra level. The above proposal will work, it's pretty simple actually, I will take care of it ASAP.

**Comment by valkenburg-prevue-ch** [comment]
You're right about slow and stuck, agreed. Ok, cool, I look forward to the feature!

**Comment by ifeulner** [comment]
This is a fundamental topic, though. This project and also terraform is meant for initial provisioning, it's actually not for operation / maintenance of the cluster and resources created by others. 
If we start taking care of such dynamically created resources (from outside, operators within the cluster, etc.), where will we stop? And does that really make sense, as there will always the context be missing. For a pure destroy of the whole cluster it could make sense... but also with the risk of deleting something that should not have been deleted.

We should be really focused.

**Comment by otavio** [comment]
I agree. I see this as an infrastructure project, and on top of it, other services are added. 

I have been doing most of the services outside the provisioning, and I only add:
* cillium
* longhorn
All other charts are added using fluxcd later.

**Comment by gthieleb** [comment]
In the current implementation this tf module does not use the helm/kubectl provider to interact directly with the K8S API and therefore is not able to manage the state. Is this unintended or a design choice?

Regarding the deletion the autoscaler FAQ states: "If a node is not needed for more than 10 minutes, it will be terminated." This can be configured see [1] for more.

I also see @ifeulner point here not to loose focus.

[1] https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-does-scale-down-work

**Comment by mysticaltech** [comment]
@gthieleb I missed your comment. So basically, for helm we use the rancher helm controller that ships with k3s, so that is the easiest thing to do and works well, hence no need for the helm provider, however, for the kubectl provider, I personally did not know it well, I just knew it existed.

For both providers, if you have ideas on how to solve this issue, or other general enhancement to the project, PRs are more welcome, and/or further guidance too.

**Comment by valkenburg-prevue-ch** [comment]
re @ifeulner 's comment: these auto-scale nodes are provided by a piece that is part of this project, I don't see why that should then be treated as something dynamic that is outside of the scope of this project. If that is so, creating the nodes should also be out of scope.

**Comment by ifeulner** [comment]
Actually the nodes are not created by this terraform script, but the cluster-autoscaler operator and it's configuration. This is a big difference.

**Comment by mysticaltech** [comment]
@valkenburg-prevue-ch Yep, exactly, that's why that issue is open and we will add another step to not be left with dangling nodes! 

And as I said above, this script can be modified to take care of autoscaled nodes, it's a good starting point. https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/master/scripts/cleanup.sh

**Comment by mysticaltech** [comment]
@captnCC If you can add support for autoscaled nodes, it would be awesome!

**Comment by mysticaltech** [comment]
Save this line to an alias in your bash preferences: `tmp_script=$(mktemp) && curl -sSL -o "${tmp_script}" https://raw.githubusercontent.com/kube-hetzner/terraform-hcloud-kube-hetzner/master/scripts/cleanup.sh && chmod +x "${tmp_script}" && "${tmp_script}" && rm "${tmp_script}"`

Then when terraform hangs, you can execute it in another window. It's a solution that works well for now.

![ksnip_20230326-052310](https://user-images.githubusercontent.com/518555/227753833-0a010931-d069-4e65-b9c9-167f23302e32.png)

![ksnip_20230326-052341](https://user-images.githubusercontent.com/518555/227753842-07f9e787-6dda-44e2-aa95-ca3e8fa3dd37.png)

**Comment by mysticaltech** [comment]
The command can be saved as an alias in your shells, as such:

```sh
alias cleanup_kh='tmp_script=$(mktemp) && curl -sSL -o "${tmp_script}" https://raw.githubusercontent.com/kube-hetzner/terraform-hcloud-kube-hetzner/master/scripts/cleanup.sh && chmod +x "${tmp_script}" && "${tmp_script}" && rm "${tmp_script}"'
```

![ksnip_20230326-074445](https://user-images.githubusercontent.com/518555/227757749-1f7dfded-3286-4e18-915c-43820bf5d2c5.png)
